feat(4144): per-dapp-selected-networks

EtherWizard33 opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

**Issue**: All dapps are connected on the same chain, if one switches, others as well. Most likely the user would prefer to have his selected networks to be specific to each dapp (aka per-dapp).

**Solution**: Add per-dapp-selected-network, Add a touchable favicon to the permission summary header that enables network switching for dapps. 


# Per-DApp Selected Network Implementation

## Key Changes

### Features
- Implemented per-dapp network selection functionality with interactive favicon
- Added network badge and switching capability in permission summary header
- Enhanced chain fallback logic for per-dapp-selected-network feature
- Added origin-specific network info for transaction reviews
- Removed non-permitted network flow checks for dapps (TODO: keep behind feature flag, enabled when !perDapp)

### Fixes
- Updated transaction notifications to use chain-specific ticker symbols
- Fixed network client ID handling for smart transactions
- Improved chain comparison logic for per-dapp transactions
- Corrected domain logo container placement behind feature flag
- Fixed favicon display logic in permission summary for unconnected dapps

### Refactoring
- Centralized per-DApp network logic using `isPerDappSelectedNetworkEnabled` helper
- Cleaned up network selection implementation
- Removed temporary network picker components
- Updated network name and image source to be origin-aware
- Standardized feature flag usage across components

### Code Quality
- Removed debug logging statements
- Code cleanup and linting improvements
- Updated test snapshots for network name changes

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Contributes to: https://github.com/MetaMask/MetaMask-planning/issues/4104

## **Manual testing steps**

Testing:
1. Enable feature flags below if they are not enabled already
2. Connect to a dapp and verify network picker appears in permissions view
3. Switch networks and verify correct network displays for that specific dapp
4. Verify network switching works properly and modal dismisses correctly
5. Verify network info displays correctly in permissions view


Make sure to include all env vars:
```
export MM_MULTICHAIN_V1_ENABLED="true"
export MM_CHAIN_PERMISSIONS="true"
export MM_PER_DAPP_SELECTED_NETWORK="true"
export MULTICHAIN_V1="true"
```

## **Screenshots/Recordings**

| Before       | After        |
|--------------|--------------|
| <img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/user-attachments/assets/7627f63c-97a6-4059-89a2-60cac1efa099"> |<img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/user-attachments/assets/62bc4dd6-d17a-4030-9c10-89df0d4f87a9"> |

<!-- [screenshots/recordings] -->


## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

EtherWizard33 committed:
feat: Add a temporary picker to test if per-dapp-selected-networks feature can work when its feature flags are enabled

make sure to include all env vars:
export MM_MULTICHAIN_V1_ENABLED="true"
export MM_CHAIN_PERMISSIONS="true"
export MM_PER_DAPP_SELECTED_NETWORK="1"
export MULTICHAIN_V1="true"

salimtb committed:
fix: fix permissisons network switch

EtherWizard33 committed:
style: remove console.logs

EtherWizard33 committed:
style: remove console.log from permission summary

EtherWizard33 committed:
style: remove console.log from network selector

EtherWizard33 committed:
feat: remove console.log from chain utils

EtherWizard33 committed:
refactor(network): use isPerDappSelectedNetworkEnabled helper

- Replace direct checks of process.env.MM_PER_DAPP_SELECTED_NETWORK in AccountPermissionsConnected.tsx with the isPerDappSelectedNetworkEnabled function.
- Update ethereum-chain-utils.js to use isPerDappSelectedNetworkEnabled when switching networks.
- Modify selectedNetworkController selectors to use isPerDappSelectedNetworkEnabled for feature flag checks.
- Add the isPerDappSelectedNetworkEnabled utility in util/networks/index.js.

This change centralizes the per-DApp network logic for better consistency and maintainability.

EtherWizard33 committed:
feat(permissions): add interactive favicon with network switching

Add a touchable favicon to the permission summary header that enables network
switching for dapps. This combines the dapp's identity (favicon/token) with
network selection in a single interactive component.

- Replace static WebsiteIcon with touchable BadgeWrapper pattern
- Add network badge to indicate current network
- Enable network switching via network selector bottom sheet
- Add fallback to AvatarToken when favicon unavailable
- Guard implementation behind per-dapp network selection feature flag
- Preserve original WebsiteIcon when feature disabled

This matches the interaction pattern from AccountPermissionsConnected,
providing a consistent way to manage dapp-specific networks across the app.

EtherWizard33 committed:
style(lint): cleanup

EtherWizard33 committed:
feat: remove temporary network picker from AcounntPermissionsConnected component

EtherWizard33 committed:
refactor: Move domain logo container behind per-dapp network feature flag

The domain logo container view was incorrectly placed outside the per-dapp network
feature flag check. This change moves the container view inside the feature flag
condition to ensure consistent UI behavior when the feature is enabled/disabled.

- Moved View wrapper inside isPerDappSelectedNetworkEnabled() check
- Simplified conditional rendering logic
- Maintains existing functionality but with proper feature flag control

EtherWizard33 committed:
fix(permission-summary): When the dapp is not yet connected, and the permission sumamry is displayed, keep showing the WebsiteIcon rather than the AvatarFavicon.

EtherWizard33 committed:
chore: in engine.ts, make use of the local feature flag isMultichainV1Enabled rather than its 'corresponding' env variable MULTICHAIN_V1

EtherWizard33 committed:
feat: remove non-permitted network flow check for dapps

Remove the permission check for non-permitted network flows in dapps, paving the way for the per-dapp-selected-network feature. Temporary screens handling these flows are pending removal upon feature completion.

EtherWizard33 committed:
feat(AddressFrom): get network info per dapp origin

- Replace global network selectors with origin-specific useNetworkInfo hook
- Update network name and image source to be origin-aware
- Update test snapshot for network name changes

EtherWizard33 committed:
feat(TransactionReview): pass origin prop to AccountFromToInfoCard

- Add origin prop to AccountFromToInfoCard for per-dapp network info
- Add debug logs for transaction review flow

OGPoyraz committed:
Use provided chainId to determine balance in useAddressBalance hook

EtherWizard33 committed:
feat: make the chain fallback logic worth in both on and off state of the per-dapp-selected-network enabled feature

EtherWizard33 committed:
feat: ensure that when checkActiveAccountAndChainId it checks the per dapp chain if there is a hostname (if its a transaction from a dapp)

EtherWizard33 committed:
doc: log the transactionMeta for debugging purposes, currently undefined

EtherWizard33 committed:
fix: per-dapp also gets the request vs dapp chain comparison

EtherWizard33 committed:
fix: pass the network client id in hopes of fixing the stx error when confirming tx from dapp while stx is enabled

pedronfigueiredo committed:
wip

pedronfigueiredo committed:
wip

EtherWizard33 committed:
fix(TransactionNotification): use chain-specific ticker for transaction notifications

The commit updates the ticker selection to use the transaction's specific chainId
instead of the global network state, ensuring correct token symbol is displayed in the TransactionSummary when openned from the TransactionNotification for
per-dapp-selected-network transactions.

EtherWizard33 committed:
refactor: use selectEvmNetworkConfigurationsByChainId instead of selectNetworkConfigurations

- and fix lint errors

EtherWizard33 committed:
Switch TransactionDetails usage to transaction specific chainId

EtherWizard33 committed:
fix: get the ticker using a selector from the component, rather than the networkConfig passed to the util function.

EtherWizard33 commented:
Making a build for QA, as of today it includes fixes to the latest known per-dapp-selected-networks issues found in the feature branch's work in progress by QA. Fixes include transactions details not showing the right ticker, and transactions failing when the user had enabled smart transactions in his mobile wallet. And includes other fixes like disconnecting from the dapp that was not working, and fallback to other network when the permission to the current active network from the dapp is removed by the user.

https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/a5e74d1e-ee19-4b25-ba88-5a614144b633

EtherWizard33 committed:
chore: cleanup debug console.log statements

EtherWizard33 committed:
chore: remove unused imports and console.logs

EtherWizard33 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard33 committed:
chore: remove redundant ts-expect-error comments

EtherWizard33 committed:
fix(smart-public-hook-test): since the networkClientId was added to app code, it also needs to be added to the mocks in tests.

EtherWizard33 committed:
Revert "feat: remove non-permitted network flow check for dapps"

This reverts commit 4193532ea4294cf8bbc3decc9fecfbd844e70480.

EtherWizard33 committed:
feat(browser): skip network check when per-dapp network selection is enabled

This change prevents the browser tab from checking network permissions when
the per-dapp network selection feature is enabled. The permission check is to ensure the dapp doesnt find itself on a network for which it doesnt have permission to after the user switched the network from global selector. When per-dapp-selected network is enabled, the dapp stands on the network it is connected to, and doesnt have to follow the network chosen from the global selector, aka, the dapp is already on a network it has given permission to. We are keeping this just in case the per-dapp-selected-network feature is removed or toggled off, in this case, the non-permitted-network flow and this permission check would be needed. Most likely this code will be removed after a few months of per-dapp-selected network feature is proven not to need to be toggled-off.

dan437 committed:
Use chainId as a second param for selectShouldUseSmartTransaction

Signed-off-by: dan437 <80175477+dan437@users.noreply.github.com>

dan437 committed:
Use chainId when calling shouldUseSmartTransaction

Signed-off-by: dan437 <80175477+dan437@users.noreply.github.com>

dan437 committed:
Always pass a chainId into the shouldUseSmartTransaction selector

Signed-off-by: dan437 <80175477+dan437@users.noreply.github.com>

dan437 committed:
Use a globalChainId if transaction.chainId is not there on the Send Confirmation page

Signed-off-by: dan437 <80175477+dan437@users.noreply.github.com>

dan437 committed:
Update the RPC URL check for STX

Signed-off-by: dan437 <80175477+dan437@users.noreply.github.com>

dan437 committed:
Enable feature flag check for STX

Signed-off-by: dan437 <80175477+dan437@users.noreply.github.com>

EtherWizard33 committed:
fix: remove networks from allowed for dev, they should be enabled locally per dev using env variables.

Otherwise QA gets builds with these chains which are not ready to QA and they start logging bugs on features still in development.

EtherWizard33 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard33 committed:
fix: bad merge in previous commit, first fix

EtherWizard33 committed:
fix(bad-merge): from previous commit, second fix, remove duplicate import.

EtherWizard33 committed:
test(WalletActions): fix the tests that broke following the addtion of per-dapp-selected-networks.

Tests were fixed mostly by adding missing mocks.

EtherWizard33 committed:
test(WalletConnectV2): fix the tests that broke following the addtion of per-dapp-selected-networks.

Tests were fixed mostly by adding missing mocks. The rest of the changes seems mostly like code formatter changes.

EtherWizard33 committed:
test(smart-tx-constants): remove networks from allowed for dev, they should be enabled locally per dev using env variables.

Otherwise QA gets builds with these chains which are not ready to QA and then log bugs on features still in development.

EtherWizard33 committed:
test(AccountActions): fix the tests that broke following the addtion of per-dapp-selected-networks.

EtherWizard22 committed:
test(TransactionDetails): Update one test's  mocked NetworkController, and the snapshot of two other tests.

The previous mock failed because the selector could not find the expected NetworkConfiguration in state, with this fix it passes. For the snapshots, it seems the previous version contained only an Amount label, while this version contains all other labels and fields we usually see in the transaction detail screen, therefore this new snapshot does seem relevant.

EtherWizard22 committed:
test(tx-controller-init): fix the unit test to match what was expected.

This can be a permanent or temporary fix just intended for the PR before review. I did message someone from the stx team who had written that line to see if the change is expected or required a deeper dive and this PR would be adjusted as needed.

EtherWizard22 committed:
test(wc-utils): add missing mock that are now necessary.

These changes are more likely related to stx changes in this per dapp selected network PR.

EtherWizard22 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard22 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard33 commented:
I have read the CLA Document and I hereby sign the CLA

EtherWizard22 commented:
I have read the CLA Document and I hereby sign the CLA

EtherWizard22 committed:
chore: temp comment

EtherWizard22 committed:
chore: lint fixes

EtherWizard22 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

christopherferreira9 commented:
Issue found with SDK:
None of the chainChanged events are being fired after the wallet_switchEthereumChain  rpc is completed.
Reproduction Steps:
- Connect
- Call wallet_switchEthereumChain for a different chain -> Returns null as expected but no chainChanged event is fired.



I've deployed a dummy SDK dapp to help testing. For this you need to:

- open this [dapp](https://mmsdk-test-dapp.vercel.app/) on a desktop browser without having MetaMask installed
- set the event listener through the console:
```
window.ethereum.on('chainChanged', (chain) => {
    console.log('chain has changed to ', chain)
})
```

- tap connect and scan the QR Code
- follow the reproduction steps.

christopherferreira9 commented:
Issue found with WalletConnect:
WalletConnect is also not switching chains before presenting the sign request.
Steps to reproduce:
- Go to https://appkit-lab.reown.com/library/wagmi/ on your desktop browser
- Connect via QR Code
- Sign -> it correctly prompts to sign on Mainnet
- Switch chain to OP (this should not change immediately)
- Sign -> See issue. It prompts the user to sign on mainnet but it should've made the prompt to sign on OP

Video:

https://github.com/user-attachments/assets/d38c09ed-2a8a-4072-a314-dd7f0ef1591b

EtherWizard22 committed:
test: temporarily exclude file from sonar coverage on new code, we ensured none of the existing tests broke, we will circle back to add coverage for the new code paths within weeks.

EtherWizard22 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard22 committed:
fix(account-permissions): had to adapt the per-dapp change to the refactoring that was made on the permissions and connections screens.

EtherWizard22 committed:
fix: lint and also chain should be effective, meaning be per dapp when per dapp is enabled otherwise global (wallet) chain.

EtherWizard22 committed:
Merge branch 'main' into featureBranch

EtherWizard22 committed:
Merge branch 'main' into feat

EtherWizard22 committed:
feat(per-dapp-feature-flag): put per dapp selected network changes for the AccountFrom components behind feature flag.

elribonazo committed:
fix: remove constasnt from setupBridge and let originatorInfo set the Bridge url

elribonazo committed:
fix: background bridge fixes always using origin getter now + minimal unitary tests

elribonazo commented:
My last changes fix https://github.com/MetaMask/metamask-mobile/issues/15239

elribonazo commented:
I have read the CLA Document and I hereby sign the CLA

EtherWizard22 committed:
feat(per-dapp-feature-flag): put per dapp selected network changes for the TransactionNotification components behind feature flag.

EtherWizard22 committed:
feat(per-dapp-feature-flag): put per dapp selected network changes for the PermissionSummary components behind feature flag.

EtherWizard22 committed:
style: remove unnecessary line

EtherWizard22 committed:
style: removing unused line

EtherWizard22 committed:
feat(per-dapp-feature-flag): put per dapp selected network changes for the TransactionDetail components behind feature

EtherWizard22 committed:
feat(per-dapp-feature-flag): put per dapp selected network changes for the  TransactionElement components behind feature flag.

EtherWizard22 committed:
chore: cleanup unnecessary white lines and unnecessary formatting changes

EtherWizard22 committed:
feat(per-dapp-feature-flag): put per dapp selected network changes for the  useSwitchNetworks hook behind feature flag.

EtherWizard22 committed:
feat(per-dapp-feature-flag): put per dapp selected network changes for the  TransactionReview Component behind feature flag.

EtherWizard22 committed:
feat(per-dapp-feature-flag): put per dapp selected network changes for the  useAddressBalances hook behind feature flag.

EtherWizard22 committed:
style: lint

EtherWizard22 committed:
feat(per-dapp-feature-flag): put per dapp selected network changes to the RPCMethodMiddleware behind feature flag.

EtherWizard22 committed:
feat(per-dapp-feature-flag): put per dapp selected network changes to the smart-publish-hook behind feature flag.

EtherWizard22 committed:
test(Confirm): update snapshot after adding the feature flag(isPerDappSelectedNetworkEnabled) to match the state of the component so that the snapshot matches the false state of this flag that will be in main.

EtherWizard22 committed:
test(TransactionDetails): update snapshot after adding the feature flag(isPerDappSelectedNetworkEnabled) to match the state of the component so that the snapshot matches the false state of this flag that will be in main.

Previous snapshot was the the ON state of the feature flag, it can be reused once we remove the feature flag and the code in main is on ON.

EtherWizard22 committed:
Revert "fix: remove constasnt from setupBridge and let originatorInfo set the Bridge url"

This reverts commit 6967275d19d05d3b474faac36819d11c4f9b9241.

EtherWizard22 committed:
Revert "fix: background bridge fixes always using origin getter now + minimal unitary tests"

This reverts commit 2f3262044ad9ceed3a2542ed422b594cb6f056c1.

EtherWizard22 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard22 committed:
chore(sonar): since new code is behind feature flag, temporarily exclude coverage on the files containing this feature flag. Exclusions to be removed and UT coverage added after removing feature flag from codebase.

EtherWizard22 committed:
chore: provide example env variables for per-dapp-selected-network feature flag

Prithpal-Sooriya reviewed (dismissed):
This looks good from a Notifications CO side.

jiexi commented on app/components/Views/NetworkSelector/useSwitchNetworks.ts:
```diff
@@ -130,6 +131,7 @@ export function useSwitchNetworks({
           origin,
           networkConfigurationId,
         );
+        isPerDappSelectedNetworkEnabled() && dismissModal?.();
```
should `isMultichainV1Enabled()` be replaced with `isPerDappSelectedNetworkEnabled()`?

EtherWizard33 replied:
Good question, I also thought about this and wasnt sure if this flag isMultichainV1Enabled is used if and only if per dapp selected network is enabled. I beleive this flag, or the env var it wraps was originally added by @adonesky1 back in September. If we are sure its ok to use isPerDappSelectedNetworkEnabled() I would change it yes.

adonesky1 replied:
Looking through it now I think `isMultichainV1Enabled` and `isPerDappSelectedNetworkEnabled` should be consolidated. I believe `MM_PER_DAPP_SELECTED_NETWORK` and `MULTICHAIN_V1` should only be one feature flag. And, per my other comments, I'm not totally clear on why we still have these feature flags at all. But at the very least we should consolidate them.

adonesky1 replied:
We've got 3 different feature flags that I believe should all just be one:
- `MM_MULTICHAIN_V1_ENABLED`
- `MM_PER_DAPP_SELECTED_NETWORK`
- `MULTICHAIN_V1`

EtherWizard33 replied:
Thanks @adonesky1 I'm glad we had a chance to confirm this over a zoom call today and will use this comment as basis for the feature flag refactoring coming in the next PR. Marking as resolved for now.

jiexi commented on app/components/Views/NetworkSelector/useSwitchNetworks.ts:
```diff
@@ -190,6 +192,7 @@ export function useSwitchNetworks({
 
       if (domainIsConnectedDapp && isMultichainV1Enabled()) {
         SelectedNetworkController.setNetworkClientIdForDomain(origin, type);
+        isPerDappSelectedNetworkEnabled() && dismissModal?.();
```
same

EtherWizard33 replied:
ok, lets see what we decide for above https://github.com/MetaMask/metamask-mobile/pull/14539/files#r2091698690

jiexi commented on app/components/Views/BrowserTab/BrowserTab.tsx:
`checkTabPermissions` exits early if PerDappSelectedNetwork is enabled, but is only called in the first place if it's false. Seems like a noop

EtherWizard33 replied:
I'm not sure I understand, could you provide some lines.

But looking at the code, it looks like the caller and callee both check the feature flag, while only one of the two would need to. But I have added both checks just in case other callers would call it to make sure it never runs unexpectedtly.

Also, other not, this check permission code is being removed at the same time as we remove this feature flag, in other words, per dapp selected network is replacing it.

jiexi replied:
```
  const checkTabPermissions = useCallback(() => {
    if (isPerDappSelectedNetworkEnabled()) {
      return;
    }
    ...
}

...

      if (!isPerDappSelectedNetworkEnabled()) {
        checkTabPermissions();
      }
```

jiexi replied:
maybe i'm misunderstanding

EtherWizard33 replied:
I should start by saying that the check permissions feature is used only when there is no per-dapp feature, and its role is to ensure that if the network is changed by the global network selector, and all dapps switch network, that a dapp doesnt find itself on a network it has not granted permission for. When per-dapp is enabled, the dapps' chain selection are independant from each other and independant of the wallet's global chain.

In the two blocks of code you listed above, the first block, the callee is checking that if it was called to run a permission check while the per-dapp feature is enabled, to exit early and not run the permission check. 

In the second block, if per-dapp is off, it runs the permission check.

Both are checking the same thing, and in this case the flag will be checked twice, but the redundance is to ensure that the checkPermissions method isnt left unguarded. And is kept at the caller's side as well for clarity so its clear right from the call site that it won't be called. I see this might be causing more confusion than clarity and if this is the case i propose to remove the checks on the caller side and keep only the callee.

EtherWizard33 replied:
Hi @jiexi, did the answer above provide a bit more clarity, please let me know, if yes, I would mark the conversation as resolved.

EtherWizard33 replied:
sent a Loom recording with a pw in slack to explain when is check permissions expected to run, hope that makes sense, please let me know if so, and I would make this convo as resolved :)

jiexi replied:
Ahh shoot. My mistake. I understand what this is doing and what your intention is now. Disregard my previous comments as they are incorrect

jiexi commented on app/components/Views/AccountPermissions/AccountPermissions.tsx:
```diff
@@ -303,10 +307,18 @@ const AccountPermissions = (props: AccountPermissionsProps) => {
+            // For global network selection, switch the active network
+            await Engine.context.MultichainNetworkController.setActiveNetwork(
+              networkClientId,
+            );
```
don't we want to change the active network regardless of if the per dapp selected network is enabled?
@adonesky1

EtherWizard33 replied:
Until Alex responds, I can provide some context, to clarify that the dapp is intended to change only only the network by hostname, this is per dapp, it's expected to be decouple from changin the wallet's global network. This also allows to release this feature regardless of the wallet being read or not to remove the global network selector.

EtherWizard33 commented on sonar-project.properties:
```diff
@@ -15,7 +15,7 @@ sonar.exclusions=**.stories.**, e2e/**, wdio/**
 
 # Excluded project files from coverage.
-sonar.coverage.exclusions=**util/test**, **/__mocks__/**
+sonar.coverage.exclusions=**util/test**, **/__mocks__/**, **NetworkConnectMultiSelector.tsx, **TransactionNotification/index.js, **BrowserTab.tsx, **RPCMethodMiddleware.ts, **PermissionsSummary.tsx, **useSwitchNetworks.ts, **ethereum-chain-utils.js, **app/selectors/selectedNetworkController.ts, **useAddressBalance.ts, **AccountPermissions.tsx, **AddressFrom.tsx, **TransactionDetails**, **Transact ...
```
thanks @jiexi and I'm glad you asked, I will paste the commit message of the corresponding commit so the intention of this temporary change is clear, and let me know if this answers your question, I'm glad to discuss it further, I did the same for the Chains Permissions project, [as you can see in this PR](https://github.com/MetaMask/metamask-mobile/pull/13401/files#diff-43ed9d31bea2a6d518d69836bcd1a8e6bd81bf4df96c4745792c220ca5aa549c), the exclusions were removed as planned.

> chore(sonar): since new code is behind feature flag, temporarily excl…
…ude coverage on the files containing this feature flag. Exclusions to be removed and UT coverage added after removing feature flag from codebase.

EtherWizard33 replied:
assuming the above answer is satisfactory for now and resolving this comment :)

sethkfman replied:
@EtherWizard33 This code should still be applicable to unit testing. You should be able to mock the flag. This is just a lot of untested code which is my concern.

jiexi commented on app/components/Views/AccountPermissions/AccountPermissions.tsx:
```diff
@@ -276,18 +277,21 @@ const AccountPermissions = (props: AccountPermissionsProps) => {
       return;
     }
 
+    const { chainId: effectiveChainId } = isPerDappSelectedNetworkEnabled()
+    ? networkInfo
+    : { chainId: currentChainId };
+
```
```suggestion
    const chainId = isPerDappSelectedNetworkEnabled()
    ? networkInfo.chainId
    : currentChainId;

```

nit

jiexi commented on app/components/UI/TransactionElement/utils.js:
```diff
@@ -601,15 +601,13 @@ function decodeConfirmTx(args) {
+    ticker,
   } = args;
-
-  const ticker = networkConfigurationsByChainId?.[txChainId]?.nativeCurrency;
```
will this result in differing behavior when `isPerDappSelectedNetworkEnabled` is false?

EtherWizard33 replied:
it should be fine on both sides of the flags (true and false), based on the fact that it is used in the same way the caller to get the ticker based on the txChainID regardless of it being a per tx dapp or not, here https://github.com/MetaMask/metamask-mobile/blob/1f99e36c7c4e9113bce7b29c0891a894ba7a61aa/app/components/UI/TransactionElement/utils.js#L921

The general idea I believe the confirmations team is following is that the transaction and transactionMeta are the source of truth and that it can be used in both cases, per-dapp chain or global chainID.

Please let me know if this addresses this question and if so I will mark it as resolved.

adonesky1 replied:
yeah this makes sense to me as long as the ticker is set by a similar check upstream

jiexi commented on app/components/UI/TransactionElement/TransactionDetails/index.js:
```diff
@@ -117,14 +120,15 @@ const createStyles = (colors) =>
+     * Chain ID string
      */
-    chainId: PropTypes.string,
+       chainId: PropTypes.string,
```
```suggestion
     chainId: PropTypes.string,
```

EtherWizard33 replied:
done, marking as resolved

jiexi commented on app/components/UI/PermissionsSummary/PermissionsSummary.tsx:
```diff
@@ -114,12 +123,62 @@ const PermissionsSummary = ({
     const iconTitle = getHost(currentEnsName || url);
 
-    return (
+    if (!isAlreadyConnected && isPerDappSelectedNetworkEnabled()) {
```
!isAlready & isPer => WebsiteIcon
!isAlready & !isPer => WebsiteIcon
isAlready & isPer => new
isAlready & !isPer => WebsiteIcon

I think the top condition should be `isAlreadyConnected && isPerDappSelectedNetworkEnabled()` which returns the not WebsiteIcon view

EtherWizard33 replied:
done, marking as resolved, ty

EtherWizard22 committed:
refactor(conditional): for readability

EtherWizard22 committed:
refactor(PermissionSummary): simplify a conditional

EtherWizard22 committed:
style: remove white space

tommasini commented on app/components/UI/AccountFromToInfoCard/AddressFrom.tsx:
```diff
@@ -74,7 +85,13 @@ const AddressFrom = ({
   }, [accountsByChainId, internalAccounts, activeAddress, origin]);
 
-  const networkImage = useSelector(selectEvmNetworkImageSource);
+  const displayNetworkName = isPerDappSelectedNetworkEnabled() 
```
Can we address this lint warnings?

EtherWizard33 replied:
yes!

tommasini replied:
They are, they are part of those 85 warnings that appear with `yarn lint`

EtherWizard33 replied:
ah warnings, I see, was looking for errors, thanks for the clarification @tommasini I have a next PR coming with small refactorings right after this one, would it be ok to address them there, to minimize need for 6-7 re-approvals? If so, then i would mark this convo as resolved for now?

vinnyhoward reviewed (dismissed):
Great work, LGTM. Pulled down and tested as well

gambinish reviewed (dismissed):
Re-approving on behalf of Notifactions

adonesky1 commented on app/util/networks/index.js:
```diff
@@ -616,6 +616,9 @@ export const isChainPermissionsFeatureEnabled = true;
 export const isPermissionsSettingsV1Enabled =
   process.env.MM_PERMISSIONS_SETTINGS_V1_ENABLED === 'true';
 
+export const isPerDappSelectedNetworkEnabled = () =>
+  process.env.MM_PER_DAPP_SELECTED_NETWORK === 'true';
+
```
Again, why are we merging it behind feature flag(s)?

EtherWizard33 replied:
discussed in zoom meeting, will refactor feature flags in next pr, marking as resolved

adonesky1 commented:
Seeing some bugs when going to actually submit transactions on non globally selected networks: https://drive.google.com/file/d/1Fxqz6_04GandCMah2raHtjjvo_ha0RjG/view?usp=drive_link

adonesky1 commented on app/core/Engine/Engine.ts:
```diff
@@ -780,11 +784,11 @@ export class Engine {
       onPreferencesStateChange: (
         listener: ({ useRequestQueue }: { useRequestQueue: boolean }) => void,
-      ) => listener({ useRequestQueue: !!process.env.MULTICHAIN_V1 }),
+      ) => listener({ useRequestQueue: isMultichainV1Enabled() }),
```
Well I suppose not given the approach is to put this behind a feature flag. So we don't want it to be true when the feature flag is off. Again I think this PR shouldn't be using a feature flag anymore, but as long as it is it makes sense that it would use the same flag

adonesky1 commented on app/core/RPCMethods/RPCMethodMiddleware.ts:
```diff
@@ -190,10 +188,17 @@ export const checkActiveAccountAndChainId = async ({
       activeChainId = ChainId[networkType as keyof typeof ChainId];
     } else if (networkType === RPC) {
-      activeChainId = providerConfigChainId;
+      activeChainId = isPerDappSelectedNetworkEnabled() ? providerConfig.chainId: providerConfigChainId;
```
This seems wrong? `providerConfig.chainId` looks like it would be the globally selected chainId? 

Just looking at this it seems the ternary will return the same chainId either way?

EtherWizard33 replied:
yes, you are right, looking at what the selectors do, the result is the same, I think the original change had come before the selectors were refactored to include 'evm' in their name and had to confirm result was the same. I've removed the unnecessary conditional, marking as resolved.

OGPoyraz committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard22 committed:
refactor(providerConfigChainId): remove unnecessary conditional, both sides were the same

EtherWizard33 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard33 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard33 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard33 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

sethkfman reviewed (dismissed):
Approving this PR. In the future, I would not recommend the approach of excluding the files from test even if they are behind feature flags.

EtherWizard22 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard22 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard33 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard22 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th

EtherWizard33 committed:
Merge branch 'main' into feat-4144-per-dapp-updated-from-main-april-7th
